### PR TITLE
Api data model & Caching

### LINKED ISSUE #74: Rework Cache Manager
The current cache manager implementation could use some overhauling:

 - It would be nice to use redis for caching instead of postgres
 - The singleton pattern seems unnecessary
 - Related to #72

### LINKED ISSUE #81: Define dataclasses for extractor results.
Currently the individual extractors return dynamic a dictionaries which somewhat follow the expected api output structure and are combined and enriched in the `MetadataManager` class. Hence, the API return models currently leak through the manager class into the extractors.
We should define a dataclass for the extractor return values which is independent of the output model. The expected output model can then be build either in the manager class, (at a later stage) even in the endpoint implementation.

AC:
- Extractors have a uniform extraction return type and output model is contained in manager or api endpoint implementation.

Note:
- If an extractor needs to deviate from a uniform output structure (e.g. because it provides extra information about how it came to its X star rating), then it should declare its own return structure declaratively. It would then also be the managers responsibility how to deal with this extra information.

### LINKED ISSUE #98: Implement caching via FastAPI Middleware
Use https://github.com/long2ice/fastapi-cache package or similar internal construct to implement caching (in postgres)

### PR DESCRIPTION
**Pull request by MRuecklCC**
- :heavy_check_mark:  needs merge of #97 
- :heavy_check_mark:  then rebase ontop of dev

Further simplifications of the data models and reimplementation (prototype) of caching as decorator / middleware

- Resolves #98 
- Resolves #74 
- Partially Resolves #81

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -24,9 +24,9 @@ tldextract = "^3.0.2"
 "pdfminer.six" = "^20201018"
 PyPDF2 = "^1.27.12"
 aiohttp = "^3.8"
-psycopg2 = "^2.9"
-SQLAlchemy = "^1.4.21"
 beautifulsoup4 = "^4.10.0"
+SQLAlchemy = "^1.4.23"
```
Comment (MRuecklCC) [inline pyproject.toml:28]: fixme: remove sqlalchemy. It is not actually used anywhere.

File: tests/integration/api_integration_test.py [hunk 2]
```diff
@@ -37,5 +39,74 @@ def test_extract_endpoint(client):
...
+        assert output.url == "https://google.com"
+        assert isinstance(output.security, MetadataTags), "did not receive data for security extractor"
+
+        assert (
+            "https://google.com" not in client.app.cache_backend._store
+        ), "should not be cached, as it was not a complete result"
+
+        with mock.patch(
+            "features.accessibility.Accessibility._execute_api_call",
+            new=AsyncMock(side_effect=lambda website_data, session, strategy: 0.1337),
```
Comment (RMeissnerCC) [inline tests/integration/api_integration_test.py:87]: What is this strategy for?

File: src/caching/backends.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        if namespace:
+            keys = list(self._store.keys())
+            for key in keys:
+                if key.startswith(namespace):
+                    del self._store[key]
+        elif key:
+            del self._store[key]
+
+
+class DatabaseBackend(Backend):
```
Comment (RMeissnerCC) [inline src/caching/backends.py:95]: Where is this being used?

File: tests/unit/database_cache_backend_test.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+from caching.backends import DatabaseBackend
+
+
+@pytest.mark.asyncio
+async def test_backend(tmpdir):
+    backend = DatabaseBackend(url=f"sqlite:///{tmpdir}/test.db")
+    await backend.setup()
+
+    assert await backend.get("foo") is None
```
Comment (RMeissnerCC) [inline tests/unit/database_cache_backend_test.py:13]: I know foo and bar are common among IT people, but personally, I find them confusing, because they have no meaning - which is also what they are for. Personally, I would prefer something like "dummy_key" and "dummy_value" or similar. This is just my opinion and not a task for you to change. I just wanted to share my thoughts here and why I will probably refrain from using foo bar myself. Furthermore we had a customer in the past who was really confused by foobar, because he came from aviation. There, they use FUBAR = fucked up beyond all recognition.  :D

File: tests/unit/database_cache_backend_test.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    await backend.set("foo", "bar", 100)
+    assert await backend.get("foo") == "bar"
+    assert await backend.get_with_ttl("foo") == (100, "bar")
+    time.sleep(3)
```
Comment (RMeissnerCC) [inline tests/unit/database_cache_backend_test.py:17]: I presume 1s would be fine here too, right? Or does it take 3s for the cache to really show changes?

File: tests/unit/database_cache_backend_test.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+@pytest.mark.asyncio
+async def test_backend_postgres():
+    # uses the configuration from docker-compose.yaml db container
+    backend = DatabaseBackend(url="postgresql://postgres:postgres@localhost/storage")
+    try:
+        await backend.setup()
+    except OSError as e:
+        print(e)
+        pytest.skip("Failed to connect to postgres, probably container is not running. Skipping test")
```
Comment (RMeissnerCC) [inline tests/unit/database_cache_backend_test.py:29]: So basically, every time the backend cannot be set up due to OSError, this test is being skipped? A nice functionality, by the way, but this could also hide corner/edge cases. Since you only catch OSError, this should be fine, though :)

**Review by RMeissnerCC — CHANGES_REQUESTED** [review]
This is a lot :D I am partially confused, e.g., by the flow of the cache. When you are done updating the tests and code, could you please give me a short tour on how the cache works and what the different functions do?

File: src/app/api.py [hunk 7]
```diff
@@ -1,127 +1,74 @@
...
     allow_headers=["*"],
 )
+app.cache_backend = InMemoryBackend()
 
 
 @app.on_event("startup")
 async def initialize():
-    logger: logging.Logger = create_logger()
-    logger.info("Startup")
+    app.logger = create_logger()
```
[the remark this answers is not in the document]
Comment (MRuecklCC) [reply src/app/api.py:27]: Yes, essentially different components will have their own child loggers. This allows to print the origin of the log message. E.g. 

```
2022-05-30 14:36:17,502 DEBUG core.metadata_manager Dispatching extractor calls.
2022-05-30 14:36:17,502 DEBUG core.metadata_manager Extracting Advertisement.
2022-05-30 14:36:17,502 INFO core.metadata_base Working on html content: Advertisement
```

File: src/app/api.py [hunk 8]
```diff
@@ -1,127 +1,74 @@
...
-    end_time = get_utc_now()
-    out = Output(
-        url=input_data.url,
-        meta=extractor_tags,
-        exception=exception,
-        time_until_complete=end_time - starting_extraction,
-    )
+@cache(
+    expire=24 * 60 * 60 * 28,
+    key=lambda input, request, response: str(input.url) if input.splash_response is None else None,
```
[the remark this answers is not in the document]
Comment (MRuecklCC) [reply src/app/api.py:49]: Not entirely sure. I think I had to because otherwise the caching backend which directly uses the return value as key may ran into trouble  (for e.g. postgresql). I'd rather go with explicit than implicit here.

File: src/app/api.py [hunk 9]
```diff
@@ -1,127 +1,74 @@
...
-        raise HTTPException(
-            status_code=400,
-            detail={
-                MESSAGE_URL: input_data.url,
-                "meta": meta_data,
-                MESSAGE_EXCEPTION: exception,
-                "time_until_complete": end_time - starting_extraction,
-            },
-        )
+    result = await app.manager.extract(input)
```
[the remark this answers is not in the document]
Comment (MRuecklCC) [reply src/app/api.py:57]: The state that was in the manager needs to live somewhere - Otherwise we would have to setup the extractors for every request (including downloading some taglists).
Another alternative would be simply have a global variable - I have no real preference for that!

File: tests/integration/api_integration_test.py (hunk 2 above)
Comment (MRuecklCC) [reply tests/integration/api_integration_test.py:87]: strategy in general is either "mobile" or "desktop". For the mock here it does not matter  - but the lambda needs to have the same signature as the original function,

File: src/caching/backends.py (hunk 3 above)
Comment (MRuecklCC) [reply src/caching/backends.py:95]: It wasn't up to now (will have to add this with the configuration via environment variables). It is a drop in replacement for the `InMemoryBackend` that is initialized at application startup. Actually, we could simply drop the `InMemoryBackend` as sqlite  kind of provides the same functionality with `sqlite://:memory:` as database url.

File: tests/unit/database_cache_backend_test.py (hunk 4 above)
Comment (MRuecklCC) [reply tests/unit/database_cache_backend_test.py:13]: Good point - I usually use meaningful variable names wherever possible :-)

File: tests/unit/database_cache_backend_test.py (hunk 5 above)
Comment (MRuecklCC) [reply tests/unit/database_cache_backend_test.py:17]: It should - The whole test is a little wonky relying on time.sleep(). If it sleeps a little to long or to short, it would fail...

File: tests/unit/database_cache_backend_test.py (hunk 6 above)
Comment (MRuecklCC) [reply tests/unit/database_cache_backend_test.py:29]: I agree, it is mainly thought for the github actions. We could start using some python docker test utilities, where the required containers are started by the unit test framework. But that seems a little overkill just for postgres - 
For splash and lighthouse we can simply mock them which make the whole test setup a lot faster and less fragile

File: src/lib/settings.py [hunk 10]
```diff
@@ -1,13 +1,11 @@
...
 
-from lib.constants import ACCESSIBILITY, SPLASH
-
 API_PORT = 5057
 LIGHTHOUSE_API_PORT = 5058
 SPLASH_PORT = 8050
 
-SPLASH_URL = f"http://{SPLASH}:{SPLASH_PORT}"
-ACCESSIBILITY_URL = f"http://{ACCESSIBILITY}:{LIGHTHOUSE_API_PORT}"
+SPLASH_URL = f"http://127.0.0.1:{SPLASH_PORT}"
```
[the remark this answers is not in the document]
Comment (RMeissnerCC) [reply src/lib/settings.py:7]: Alright, I will leave this unresolved for now :)

[the remark this answers is not in the document]
Comment (RMeissnerCC) [reply src/lib/settings.py:7]: Alright, I will leave this unresolved for now :)

[the remark this answers is not in the document]
Comment (RMeissnerCC) [reply src/lib/settings.py:7]: Alright, I will leave this unresolved for now :)

[the remark this answers is not in the document]
Comment (RMeissnerCC) [reply src/lib/settings.py:7]: Alright, I will leave this unresolved for now :)

File: src/app/api.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (RMeissnerCC) [reply src/app/api.py:57]: Me neither. Probably this indicates, that some major refactoring would help us in the future. Fine for now.

[the remark this answers is not in the document]
Comment (RMeissnerCC) [reply src/app/api.py:57]: Me neither. Probably this indicates, that some major refactoring would help us in the future. Fine for now.

File: src/caching/backends.py (hunk 3 above)
Comment (RMeissnerCC) [reply src/caching/backends.py:95]: Then yes, please, fuse both classes.

Comment (RMeissnerCC) [reply src/caching/backends.py:95]: Then yes, please, fuse both classes.

File: tests/unit/database_cache_backend_test.py (hunk 4 above)
Comment (RMeissnerCC) [reply tests/unit/database_cache_backend_test.py:13]: Great, thank you :)

File: tests/unit/database_cache_backend_test.py (hunk 5 above)
Comment (RMeissnerCC) [reply tests/unit/database_cache_backend_test.py:17]: In an ideal world, we could mock time and just move it forward. I think in an old project we had some code on this. But it is fine for now :) Thanks for clarifying.

File: tests/unit/database_cache_backend_test.py (hunk 6 above)
Comment (RMeissnerCC) [reply tests/unit/database_cache_backend_test.py:29]: Alright, sounds good :) Should I leave this open for now, or would splash and lighthouse be mocked in the future.

Comment (MRuecklCC) [reply tests/unit/database_cache_backend_test.py:29]: Splash and lighthouse are already mocked in the integration tests :-)

File: src/app/api.py [hunk 11]
```diff
@@ -1,127 +1,74 @@
...
-                    stars=meta[key][STAR_CASE],
-                    time_for_completion=meta[key][TIME_REQUIRED],
-                    explanation=meta[key][EXPLANATION],
-                ),
-            )
-
-    return extractor_tags
+@app.middleware("http")
+async def caching_and_response_time(request: Request, call_next):
+    start_time = time.time()
```
[the remark this answers is not in the document]
Comment (MRuecklCC) [reply src/app/api.py]: i replaced all those measurements with a context manager:

```python
with runtime() as t:
      import time
      time.sleep(1)

print(f"Execution time: {t():.4f} secs")
```

much nicer imho :)

File: src/caching/backends.py (hunk 3 above)
Comment (MRuecklCC) [reply src/caching/backends.py:37]: removed the in memory part

File: tests/unit/database_cache_backend_test.py (hunk 4 above)
Comment (MRuecklCC) [reply tests/unit/database_cache_backend_test.py]: done

File: tests/integration/api_integration_test.py (hunk 2 above)
Comment (RMeissnerCC) [reply tests/integration/api_integration_test.py:106]: Sorry, my reply here from yesterday got lost: if strategy is either mobile or desktop, then it should not be a float - could you replace the float please with a suitable string?

File: src/core/metadata_manager.py [hunk 12]
```diff
@@ -84,73 +65,48 @@ def _extractors(cls) -> list[Type[MetadataBase]]:
...
+        site = await WebsiteData.from_input(
+            input=message,
+            logger=self.logger,
             tld_extractor=self.tld_extractor,
         )
 
-        self._logger.debug(f"WebsiteManager loaded at {time.perf_counter() - global_start} since start")
+        self.logger.debug("Build website data object.")
+
+        async def run_extractor(extractor: MetadataBase) -> Union[MetadataTags, Error]:
```
[the remark this answers is not in the document]
Comment (MRuecklCC) [reply src/core/metadata_manager.py:101]: Its only used in this function -> should not be part of the public api. AFAIK there are no async lambdas, hence the explicit function declaration.

The pros of having it nested: no one else can see it becomes immediately clear if it is no longer required. I like it because it prevents piling up unused code and growing coupling between files.

[the remark this answers is not in the document]
Comment (RMeissnerCC) [reply src/core/metadata_manager.py:101]: I agree with your pro points, the first one could be solved with __ as well, due to name wrangling.
Maybe I just need to get used to it :/

File: tests/integration/api_integration_test.py (hunk 2 above)
Comment (MRuecklCC) [reply tests/integration/api_integration_test.py:106]: you got something wrong here! The `_execute_api_call` function that gets mocked takes an strategy argument that must be either "desktop" or "mobile" but returns the score which is a float. 

The async mock here has to have this exact signature, but instead of issuing a request to the lighthouse container, it simply returns 0.1337 as a result allowing us test the accessibility extractor as if the container was actually running.

File: src/app/api.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (MRuecklCC) [reply src/app/api.py]: i moved them into global scope

File: src/core/metadata_manager.py [hunk 13]
```diff
@@ -30,37 +30,18 @@
...
 from lib.logger import get_logger
-from lib.timing import get_utc_now, global_start
 
 
 class MetadataManager:
-    def __init__(self, extractors: list[MetadataBase]) -> None:
-        self._logger = get_logger()
-        self.metadata_extractors = extractors
+    def __init__(self, extractors: list[MetadataBase]):
+        self.logger = get_logger()
```
[the remark this answers is not in the document]
Comment (MRuecklCC) [reply src/core/metadata_manager.py]: Having multiple loggers for different  components of the app allows to have the module name automatically added to the log messages. E.g.

```
2022-05-31 11:46:26 INFO core.metadata_manager Extracted ExtractFromFiles in  1.09s.
2022-05-31 11:46:26 INFO core.metadata_manager Extracted Accessibility in  0.00s.
2022-05-31 11:46:26 DEBUG core.metadata_manager Received all extractor results.
```
Where the `logger = logging.getLogger(__name__)` results in the nice `core.metadata_manager` or other module names. We could explicitly pass loggers around via constructors, but then the manager would log messages under `app.api`.

File: src/caching/cache.py [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+            if cache_key is None and only_if_cached:
+                logger.debug("Request cannot be served from cache")
+                # todo: discuss: This is somewhat questionable. If the user requests a response served
+                #       from cache, but the response is not cached - and can never be, what should we do?
+                raise HTTPException(status_code=400, detail="Request never be served from cache.")
+            elif cache_key is None:
+                logger.debug("Caching not supported for request")
+                return await func(*args, **kwargs)
+
+            age, ret = await backend.get_with_ttl(cache_key)
```
[the remark this answers is not in the document]
Comment (MRuecklCC) [reply src/caching/cache.py:81]: `get_with_ttl` returns a tuple of age and value, the value here gets put into the ret variable, because it will eventually be returned

File: src/caching/cache.py [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+                logger.info(f"Caching disabled by Cache-Control: {request.headers.get('Cache-Control')}")
+                return await func(*args, **kwargs)
+
+            cache_key = key(**kwargs)
+
+            if cache_key is None and only_if_cached:
+                logger.debug("Request cannot be served from cache")
+                # todo: discuss: This is somewhat questionable. If the user requests a response served
+                #       from cache, but the response is not cached - and can never be, what should we do?
+                raise HTTPException(status_code=400, detail="Request never be served from cache.")
```
[the remark this answers is not in the document]
Comment (MRuecklCC) [reply src/caching/cache.py:76]: absolutely, fixed!

File: src/caching/cache.py [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+    :param expire: How long (seconds) a stored cache entry will be valid.
+    :param backend: The cache storage engine to be used.
+    :param key: A function that defines whether an incoming request can be cached and under which key it will be cached.
+                It will take the same arguments as the decorated function and should return a string key if the request
+                can be cached, or None otherwise.
+    :return: A function that just looks like the decorated function but is wrapped with caching behaviour.
+    """
+
+    def wrapper(func):
+        # Note:
```
[the remark this answers is not in the document]
Comment (MRuecklCC) [reply src/caching/cache.py:36]: Took me some time to figure that out... So I guess its good to note that down :D

File: src/caching/backends.py [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+        return r if r is None else r["value"]
+
+    async def set(self, key: str, value: str, expire: int = None):
+        # upsert data, works for postgres and sqlite databases.
+        await self.database.execute(
+            query="""
+            INSERT INTO metadata_cache (key, ttl, value) VALUES (:key, :ttl, :value)
+             ON CONFLICT (key) DO UPDATE SET ttl = EXCLUDED.ttl, value=EXCLUDED.value
+             """,
+            values={"key": key, "ttl": self.now() + (expire or 0), "value": value},
```
[the remark this answers is not in the document]
Comment (MRuecklCC) [reply src/caching/backends.py:78]: Mhhh, I don't know, it really makes things less readable imho... Whereas when refactoring, a search replace in that file should mainly do it. If you insist, I will ad them :P

[the remark this answers is not in the document]
Comment (MRuecklCC) [reply src/caching/backends.py:78]: E.g. he query would read: 

```python
 f"""
            INSERT INTO metadata_cache ({KEY}, {TTL}, {VALUE}) VALUES (:{KEY}, :{TTL}, :{VALUE})
             ON CONFLICT ({KEY}) DO UPDATE SET {TTL} = EXCLUDED.{TTL}, {VALUE}=EXCLUDED.{VALUE}
             """
```

[the remark this answers is not in the document]
Comment (RMeissnerCC) [reply src/caching/backends.py:78]: I did not mean the query, I meant the key and ttl in the python code. In the query I do not mind.

File: src/caching/backends.py [hunk 18]
```diff
@@ -0,0 +1,141 @@
...
+
+    @property
+    def _now(self) -> int:
+        return int(time.time())
+
+    def _get(self, key: str):
+        v = self._store.get(key)
+        if v:
+            if v.ttl_ts < self._now:
+                del self._store[key]
```
[the remark this answers is not in the document]
Comment (MRuecklCC) [reply src/caching/backends.py]: Mhh, no clue, this was the stolen part from the open source library. I removed the whole InMemoryBackend class

File: src/caching/cache.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (RMeissnerCC) [reply src/caching/cache.py:81]: If it returns a value, why is it "ret"? And what is ttl (I guess time to live)? I prefer to spell out variable names, so they explicit and I do not have to guess.

File: src/caching/backends.py [hunk 19]
```diff
@@ -0,0 +1,141 @@
...
+
+    @abc.abstractmethod
+    async def clear(self, key: str = None):
+        pass
+
+
+@dataclass
+class Value:
+    data: str
+    ttl_ts: int
```
[the remark this answers is not in the document]
Comment (MRuecklCC) [reply src/caching/backends.py]: That was the helper for the InMemoryBackend below, obsolete and removed now

File: src/caching/backends.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (MRuecklCC) [reply src/caching/backends.py:82]: yeah, but when i only use it in the dict, then the whole refactoring idea is broken...

File: src/caching/cache.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (MRuecklCC) [reply src/caching/cache.py]: renamed ret to result - ttl is a pretty standard short form when it comes to caching,

File: src/caching/backends.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (RMeissnerCC) [reply src/caching/backends.py:82]: Mmmh, true, too. I presume using the .insert functionality of the database package, like I used for the timeline in metaqs, cannot be used here? This way you have less explicit SQL code.

File: src/caching/cache.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (RMeissnerCC) [reply src/caching/cache.py]: Ok

File: src/core/metadata_manager.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (RMeissnerCC) [reply src/core/metadata_manager.py]: Alright :)

File: src/caching/backends.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (MRuecklCC) [reply src/caching/backends.py:82]: I think that functionality comes from sqlalchemy - I would keep it simple stupid here. There is one table with three columns, pulling in some ORM or query builder seems overkill to me :-/. Also the whole point of the Backend class is to hide the SQL behind an abstraction layer and keep it out of the caching decorator.

[the remark this answers is not in the document]
Comment (RMeissnerCC) [reply src/caching/backends.py:82]: Alright, let us keep this as is. :)

**Review by RMeissnerCC — APPROVED** [review]
Thanks for the nice changes.
